api/src/routes/datasets.js

```diff
@@ -1074,27 +1133,39 @@ router.post(
             },
           });
         } catch (e) {
+          logger.error('Error creating stage request log', e);
+          return next(createError(500, 'Error creating stage request log'));
           // console.log()
         }
       }
-      const wf_name = req.params.wf;
+      console.log(`Starting workflow ${wf_name} on dataset ${dataset.id}`);
       const wf = await datasetService.create_workflow(dataset, wf_name, req.user.id);
       return res.json(wf);
     }
 
-    if (req.params.wf === CONSTANTS.WORKFLOWS.PROCESS_DATASET_UPLOAD
-          || req.params.wf === CONSTANTS.WORKFLOWS.CANCEL_DATASET_UPLOAD) {
-      // return 403 Forbidden if the user is not allowed to initiate the requested workflow.
-      verifyUploadEnabledForRole(req, res, next);
-
-      const { workflowInitiated, workflowInitiationError } = await initiateUploadWorkflow({
-        dataset,
-        requestedWorkflow: req.params.wf,
-        user: req.user,
+    if (wf_name === CONSTANTS.WORKFLOWS.PROCESS_DATASET_UPLOAD
+          || wf_name === CONSTANTS.WORKFLOWS.CANCEL_DATASET_UPLOAD) {
+      verifyUploadEnabledForRole(req, res, (err) => {
+        if (err) {
+          return next(err);
+        }
+        // Continue with the workflow initiation
+        initiateUploadWorkflow({
+          dataset,
+          requestedWorkflow: wf_name,
+          user: req.user,
+        }).then(({ workflowInitiated, workflowInitiationError }) => {
+          if (workflowInitiated) {
+            return res.json(workflowInitiated);
+          }
+          next(createError.InternalServerError(workflowInitiationError));
+        }).catch(next);
```

Comment:
forward to next middleware based upon successful workflow initiation or upon error to ensure proper status code are returned instead of always returning 500.